examples/tutorials/core_api_mnist/model_def_adaptive.py

```diff
@@ -0,0 +1,188 @@
+    parser.add_argument('--lr', type=float, default=1.0, metavar='LR',
+                        help='learning rate (default: 1.0)')
+    parser.add_argument('--gamma', type=float, default=0.7, metavar='M',
+                        help='Learning rate step gamma (default: 0.7)')
+    parser.add_argument('--no-cuda', action='store_true', default=False,
+                        help='disables CUDA training')
+    parser.add_argument('--no-mps', action='store_true', default=True,
+                        help='disables macOS GPU training')
+    parser.add_argument('--dry-run', action='store_true', default=False,
+                        help='quickly check a single pass')                    
+    parser.add_argument('--seed', type=int, default=1, metavar='S',
+                        help='random seed (default: 1)')
+    parser.add_argument('--log-interval', type=int, default=10, metavar='N',
+                        help='how many batches to wait before logging training status')
+    parser.add_argument('--save-model', action='store_true', default=True,
+                        help='For Saving the current Model')
+  
+    args = parser.parse_args()
+    use_cuda = not args.no_cuda and torch.cuda.is_available()
+    use_mps = not args.no_mps and torch.backends.mps.is_available()
+
+    info = det.get_cluster_info()
+    assert info is not None, "this example only runs on-cluster"
+    latest_checkpoint = info.latest_checkpoint
+    if latest_checkpoint is not None: 
+        with core_context.checkpoint.restore_path(latest_checkpoint) as path:
+            model = load_state(path)
+
+    # NEW: get hparams from config
+    hparams = info.trial.hparams
+
+    torch.manual_seed(args.seed)
+
+    if use_cuda:
+        device = torch.device("cuda")
+    elif use_mps:
+        device = torch.device("mps")
+    else:
+        device = torch.device("cpu")
+
+    train_kwargs = {'batch_size': args.batch_size}
+    test_kwargs = {'batch_size': args.test_batch_size}
+    if use_cuda:
+        cuda_kwargs = {'num_workers': 1,
+                       'pin_memory': True,
+                       'shuffle': True}
+        train_kwargs.update(cuda_kwargs)
+        test_kwargs.update(cuda_kwargs)
+
+    transform=transforms.Compose([
+        transforms.ToTensor(),
+        transforms.Normalize((0.1307,), (0.3081,))
+        ])
+    dataset1 = datasets.MNIST('../data', train=True, download=True,
+                       transform=transform)
+    dataset2 = datasets.MNIST('../data', train=False,
+                       transform=transform)
+    train_loader = torch.utils.data.DataLoader(dataset1,**train_kwargs)
+    test_loader = torch.utils.data.DataLoader(dataset2, **test_kwargs)
+
+    # NEW: pass relevant hparams to model and optimizer
+    model = Net(hparams).to(device)
+    optimizer = optim.Adadelta(model.parameters(), lr=hparams["learning_rate"])
+
+    scheduler = StepLR(optimizer, step_size=1, gamma=args.gamma)
+    for epoch in range(1, args.epochs + 1):
```

Comment:
There's an issue here where you can't respect both `args.epochs` and the hyperparameter search.

And adaptive asha search is going to try a very large number of hyperparameter samples for a small amount of training, promote the best half or quarter of those, then train those for 2x or 4x as long, then repeat that process until there's only one best trial training for much longer than all the others.

(It's slightly more complex than that, but that's a good mental model for how it works)

Adaptive asha is a good hpsearch algorithm because of theories like the [Lottery Ticket Hypothesis](https://arxiv.org/abs/1803.03635v1) which says that one of the most important factors to a model is how it gets initialized.  So by trying a very large number of trials for a very short time, the chances that you get to pick a "naturally talented model" is pretty high.  And since you don't train them for very long, you don't waste very much time in the process.

But this training script will always train for `args.epochs` epochs; it won't respect the differing lengths that the adaptive asha algorithm wants it to train for.  That is going to effective turn your Adaptive Asha algorithm into a glorified random searcher, where you just pick a bunch of random hyperparameters and train them for the same length.  Asha will pick a _lot_ of hyperparameter samples and they'll all train to completion and it will take a long time.

You're going to need to do the `for op in core_context.searcher.operations()` pattern like in the `3_hpsearch.py` of the integer incrementing core api tutorial.

Reply:
Thank you for the explanation! Did not realize that